core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/CostBasedScheduleGenerator.scala

```diff
@@ -82,71 +90,187 @@ class CostBasedScheduleGenerator(
+            outputPortIdsToViewResult
+
+        // allocate an URI for each of these output ports
+        val outputPortConfigs: Map[GlobalPortIdentity, OutputPortConfig] =
+          outputPortIdsNeedingStorage.map { gpid =>
+            val outputWriterURI = createResultURI(
+              workflowId = workflowContext.workflowId,
+              executionId = workflowContext.executionId,
+              globalPortId = gpid
             )
-          ) ++ portIdsToViewResult
-        val portConfigs = portIdsNeedingStorage
-          .map(outputPortId =>
-            outputPortId -> {
-              val uri = createResultURI(
-                workflowId = workflowContext.workflowId,
-                executionId = workflowContext.executionId,
-                globalPortId = outputPortId
-              )
-              PortConfig(storageURI = uri)
-            }
-          )
-          .toMap
-        val resourceConfig = ResourceConfig(portConfigs = portConfigs)
-        val ports = operators.flatMap(op =>
+            gpid -> OutputPortConfig(outputWriterURI)
+          }.toMap
+
+        val resourceConfig = ResourceConfig(portConfigs = outputPortConfigs)
+
+        // --- enumerate all ports belonging to the Region ------------------
+        val ports: Set[GlobalPortIdentity] = physicalOps.flatMap { op =>
           op.inputPorts.keys
             .map(inputPortId => GlobalPortIdentity(op.id, inputPortId, input = true))
             .toSet ++ op.outputPorts.keys
             .map(outputPortId => GlobalPortIdentity(op.id, outputPortId))
             .toSet
-        )
+        }
+
+        // build the Region skeleton (no input‑port URIs yet)
         Region(
           id = RegionIdentity(idx),
-          physicalOps = operators,
+          physicalOps = physicalOps,
           physicalLinks = links,
           ports = ports,
           resourceConfig = Some(resourceConfig)
         )
     }
+
+    // Collect writer‑side configs so we can look them up in Pass 2
+    val allOutputPortConfigs: Map[GlobalPortIdentity, OutputPortConfig] =
+      regionsWithOnlyOutputPortURIs
+        .flatMap(_.resourceConfig) // Seq[ResourceConfig]
+        .flatMap(_.portConfigs.collect { // PortConfig → OutputPortConfig
+          case (id, cfg: OutputPortConfig) => id -> cfg
+        })
+        .toMap
+
+    /* ------------------------------------------------------------
+     * Pass 2 – add input‑port storage configs (reader URIs)
+     * ------------------------------------------------------------ */
+
+    regionsWithOnlyOutputPortURIs.map { existingRegion =>
+      val nonDepMatEdges: Set[PhysicalLink] = matEdges.diff(physicalPlan.getDependeeLinks)
+
+      /**
+        * MatEdges that originally connected to the input ports of this region.
+        */
+      val relevantMatEdges: Set[PhysicalLink] = nonDepMatEdges.filter { matEdge =>
+        existingRegion.getOperators.exists(_.id == matEdge.toOpId)
```

Comment:
try to use block comment (`/** xxx */`) only for methods or a comment with a long paragraph. use line comment (`// xxx`) for a single line.